### PR TITLE
[core] Spill tasks from waiting queue

### LINKED ISSUE #14094: [core] Tasks are not spilled back when waiting for dependencies, even when there are remote resources available
### What is the problem?

I just found it while I was looking at code. Probably not a high priority to fix. Created an issue for tracking purposes.

It looks like if tasks are queued because of admission control, they are not spilled back although there are resources in other nodes. I think it is expected because we don't take into account of the object store memory when we spillback tasks.

This can be solved if we can fill up the object-store memory usage of each task automatically (I think this is possible since when we pull, we'd know the object sizes of all dependencies). 

Imagine this scenario.

1. There are 3 nodes A, B, and C.
2. Let's imagine there's a big object created at node C by a task. (80MB dependency)
3. Imagine node A has 4 cpus and 100MB object store memory. and node B has 8 cpus with 1000MB object store memory.
```
nodeA: {CPU:0/4, object_store_memory: 0/100MB}
nodeB: {CPU: 0/8, object_store_memory: 0/1000MB}
```
4. Now, create two large dependencies, 80MB each at node C. Let's call it dep1 and dep2.
5. Schedule the first long-running task with dep1 to node A. Note that the object store memory of node A is now filled, and it cannot pull any more objects by admission control.
```
nodeA: {CPU:3/4, object_store_memory: 80MB/100MB}
nodeB: {CPU: 0/8, object_store_memory: 0/1000MB}
```
6. When we schedule long-running tasks with dep2, they are supposed to be scheduled on node B because node B has enough resources. (It has plenty of CPUs and object store memory to schedule tasks with dep2, 80MB).
7. With the reproduction script underneath, you can verify they are not scheduled. Tasks are not spilled back from the head node. 
8. You can verify tasks are correctly spilled back if there's no admission control involved (for example, increase the head node object store memory to 1000MB).

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

- run the script with a node a object store memory 100MB. You can see only 1 CPU - 5 CPUs is utilized (idk when 5 CPUs are utilized. Most of the time, only 1 CPU is utilized).
- If you increase the memory to 1000MB in a head node, 9 CPUs are utilized (expected behavior).

```python3
from ray.cluster_utils import Cluster
import ray
import numpy as np
import time

cluster = Cluster()

# Add a head node. This will work correctly if the object store memory is 1000MB. Not working if it is 100MB.
node_a = cluster.add_node(num_cpus=4, object_store_memory=100 * 1024 * 1024, resources={"head": 1.0})
cluster.wait_for_nodes()
ray.init(address=cluster.address)

# This worker node should have enough memory to run many tasks
node_b = cluster.add_node(num_cpus=8, object_store_memory=1000*1024*1024)
# This worker node will create a large dependency that is pulled to the head node. It won't do anything other than that.
node_c = cluster.add_node(num_cpus=0, object_store_memory=1000*1024*1024, resources={"worker_for_object":1.0})
cluster.wait_for_nodes()

@ray.remote(resources={"head":0.1})
def task_with_big_object_dep(dep):
    import time
    print(f"dep1 task is scheduled on {ray.get_runtime_context().node_id}")
    time.sleep(30)

@ray.remote
def task_with_big_object_dep2(dep):
    import time
    print(f"dep2 task is scheduled on {ray.get_runtime_context().node_id}")
    time.sleep(30)

@ray.remote(num_cpus=0, resources={"worker_for_object":1.0})
def create_dep_in_worker():
    arr = np.zeros(80 * 1024 * 1024, dtype=np.uint8)  # 80 MB
    return arr

dep = create_dep_in_worker.remote()
dep2 = create_dep_in_worker.remote()

# Head node {CPU:1/4, object_store_memory:80 MB/100 MB}
# Worker node {CPU:0/8, object_store_memory:0/1000 MB}
head_node_task = task_with_big_object_dep.remote(dep)
time.sleep(3)

# Now schedule tasks with the second dependency.
# These tasks should be scheduled in the node B because it has tons of memory & CPUs
# Ideal status:
# Head node {CPU:1/4, object_store_memory:80 MB/100 MB} -> Cannot schedule more task because the arg is pinned in memory.
# worker node: {CPU: 8/8, object_store_memory: 640 MB / 1000 MB}
# This means the total cpu allocation should be 9 / 12
worker_node_tasks = [task_with_big_object_dep2.remote(dep2) for _ in range(8)]

# See the resource usages.
print(f"Head node node id: {node_a.raylet_socket_name}")
print(f"Worker node node id: {node_b.raylet_socket_name}")
while True:
    time.sleep(1)
    print(f"CPU usages: {ray.cluster_resources()['CPU'] - ray.available_resources()['CPU']}/{ray.cluster_resources()['CPU']}")
```

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Take 2 on #14094. The issue is that we do not spill tasks from the waiting queue. Also, object store memory is treated differently from other resources, so the current scheduling policy is not enough. There are a few reasons for this:
1. Object store memory can be shared by multiple tasks that have overlapping dependencies.
2. Object store memory can be used in other ways, e.g., for pinned task outputs.
3. Object store memory requirements are computed from the task argument size, not requested explicitly by the user.

To compensate, this PR checks whether a task's dependencies are being fetched or not, to determine whether there is enough object store memory locally.
```
Previous scheduling policy:
  Resources available locally -> do not spill
  Resources not available locally -> spill from the dispatch queue (tasks with arguments already local)
New scheduling policy:
  Resources available locally and deps are being fetched for waiting tasks -> do not spill
  Resources available locally and deps are not being fetched for waiting tasks -> spill
  Resources not available locally -> spill
```

## Related issue number

Closes #14094.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/object_manager/object_manager.h [hunk 1]
```diff
@@ -198,6 +199,10 @@ class ObjectManager : public ObjectManagerInterface,
   /// Get the port of the object manager rpc server.
   int GetServerPort() const { return object_manager_server_.GetPort(); }
 
+  bool IsPullRequestInactiveDueToOom(uint64_t pull_request_id) const {
```
Comment (ericl) [inline src/ray/object_manager/object_manager.h:202]:
```suggestion
  bool NotBeingPulledYet(uint64_t pull_request_id) const {
```

Comment (ericl) [reply src/ray/object_manager/object_manager.h:202]: Actually would suggest inverting this: `PullActive`

File: src/ray/object_manager/pull_manager.cc [hunk 2]
```diff
@@ -162,6 +162,18 @@ void PullManager::UpdatePullsBasedOnAvailableMemory(size_t num_bytes_available)
...
+  if (num_bytes_being_pulled_ >= num_bytes_available_) {
+    // We have at least as many pull requests queued than we do memory
+    // capacity.
+    at_memory_capacity_ = true;
+  } else {
+    // We can potentially serve another pull request. There may actually be
+    // more requests in the queue, since we do not serve requests until we have
+    // the object size information.
+    at_memory_capacity_ = false;
+  }
```
Comment (ericl) [inline src/ray/object_manager/pull_manager.cc:175]: We can simplify by not checking this. I think it's fine to spill if we don't have object capacity information yet.

**Comment by ericl** [comment]
Main comment is to remove the OOM check condition.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.h [hunk 3]
```diff
@@ -156,7 +160,7 @@ class ClusterResourceScheduler : public ClusterResourceSchedulerInterface {
   //           task request.
   std::string GetBestSchedulableNode(
       const std::unordered_map<std::string, double> &task_request, bool actor_creation,
-      int64_t *violations, bool *is_infeasible);
+      bool force_spillback, int64_t *violations, bool *is_infeasible);
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.h:163]: Why don't we just make it have a default argument false?

**Comment by stephanie-wang** [comment]
> Main comment is to remove the OOM check condition.

We need the OOM check to let the scheduling policy know whether we should prefer local scheduling or not. We can sync offline if this is still unclear.

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 4]
```diff
@@ -887,6 +895,61 @@ bool ClusterTaskManager::ReturnCpuResourcesToBlockedWorker(
 void ClusterTaskManager::ScheduleAndDispatchTasks() {
   SchedulePendingTasks();
   DispatchScheduledTasksToWorkers(worker_pool_, leased_workers_);
+  // TODO(swang): Spill from waiting queue first? Otherwise, we may end up
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:898]: This sounds like a good idea? Why do you not do that here? Any edge case in your mind?

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 5]
```diff
@@ -887,6 +895,61 @@ bool ClusterTaskManager::ReturnCpuResourcesToBlockedWorker(
...
+  // spilling a task whose args are already local.
+  SpillWaitingTasks();
+}
+
+void ClusterTaskManager::SpillWaitingTasks() {
+  // Try to spill waiting tasks to a remote node, prioritizing those at the end
+  // of the queue. Waiting tasks are spilled if there are enough remote
+  // resources AND (we have no resources available locally or we cannot fetch
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:906]:
```suggestion
  // resources AND (we have no resources available locally OR we cannot fetch
```

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 6]
```diff
@@ -887,6 +895,61 @@ bool ClusterTaskManager::ReturnCpuResourcesToBlockedWorker(
...
+    // feasible node, even if we have enough resources available locally for
+    // placement.
+    bool out_of_memory =
+        task_dependency_manager_.TaskDependenciesBlockedDueToOom(task_id);
+    auto placement_resources =
+        task.GetTaskSpecification().GetRequiredPlacementResources().GetResourceMap();
+    int64_t _unused;
+    bool is_infeasible;
+    // TODO(swang): The policy currently does not account for object store
+    // memory availability. Ideally, we should pick the node with the most
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:932]: Isn't it causing flooding? I think it'd be okay to just do random spilling like now

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 7]
```diff
@@ -887,6 +895,61 @@ bool ClusterTaskManager::ReturnCpuResourcesToBlockedWorker(
...
+  // resources AND (we have no resources available locally or we cannot fetch
+  // their dependencies due to lack of object store memory). We consider object
+  // store memory separately from the task's placement resources since object
+  // store memory is managed separately from other resources.
+  //
+  // NOTE(swang): We do not iterate by scheduling class here, so if we break
+  // due to lack of remote resources, it is possible that a waiting task that
+  // is earlier in the queue could have been scheduled to a remote node.
+  auto it = waiting_task_queue_.end();
+  while (it != waiting_task_queue_.begin()) {
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:915]: Can we specify the end of the queue is not the same as pulling order?

File: src/ray/raylet/scheduling/cluster_task_manager_test.cc [hunk 8]
```diff
@@ -954,6 +960,94 @@ TEST_F(ClusterTaskManagerTest, RleaseAndReturnWorkerCpuResources) {
...
+  task_manager_.ScheduleAndDispatchTasks();
+  ASSERT_EQ(num_callbacks, 0);
+
+  // All waiting tasks are blocked due to lack of memory. We should only spill
+  // up to the remote node's resource availability.
+  for (auto &task : tasks) {
+    dependency_manager_.blocked_tasks.insert(task.GetTaskSpecification().TaskId());
+  }
+  task_manager_.ScheduleAndDispatchTasks();
+  ASSERT_EQ(num_callbacks, 2);
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager_test.cc:1002]: Nice test!

**Review by rkooo567 — APPROVED** [review]
LGTM.

**Comment by stephanie-wang** [comment]
Synced with @ericl offline and we decided on these changes:
- check if pull request is active or not instead of checking at memory capacity. Will also add a check to make sure that the pull request has all its object size information, since this is also a reason why it could be inactive.
- add a Python test to make sure we are triggering spillback correctly in the case that there are enough CPUs but not enough memory.
- invoke the scheduler periodically to handle the above case.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.h (hunk 3 above)
Comment (stephanie-wang) [reply src/ray/raylet/scheduling/cluster_resource_scheduler.h:163]: I think it's better to be explicit.

File: src/ray/raylet/scheduling/cluster_task_manager.cc (hunk 4 above)
Comment (stephanie-wang) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:898]: It's a bit hard to reason about. I'd rather do this if there is actually a bug associated with it.

**Comment by stephanie-wang** [comment]
FYI, I added a regression test for spilling back on OOM but there is a bug in the resource reporting that is actually causing resources to get reported on every heartbeat, so the test is actually passing. I added a TODO to trigger spillback from the waiting queue instead of doing it in this PR.

**Comment by stephanie-wang** [comment]
Note: Need to fix tests.

**Review by ericl — APPROVED** [review]

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 9]
```diff
@@ -887,6 +895,62 @@ bool ClusterTaskManager::ReturnCpuResourcesToBlockedWorker(
...
+    it--;
+    const auto &task = std::get<0>(*it);
+    const auto &task_id = task.GetTaskSpecification().TaskId();
+    // Check whether this task's dependencies are not being pulled.  If this is
+    // true, then we should force the task onto a remote feasible node, even if
+    // we have enough resources available locally for placement.
+    bool force_spillback = !task_dependency_manager_.TaskDependenciesPending(task_id);
```
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:925]:
```suggestion
    bool force_spillback = task_dependency_manager_.CannotExecuteLocallyYet(task_id);
```

Comment (ericl) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:925]: (avoid double negative)

Comment (stephanie-wang) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:925]: I don't think this name is accurate. The call just returns whether dependencies are being fetched, not whether the dependencies are local yet.

Comment (ericl) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:925]: (also avoid "pending" since arguably everything is "pending" in the task dep manager)

Comment (stephanie-wang) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:925]: How about "TaskDependenciesBlocked" then?

Comment (ericl) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:925]: :+1: blocked

**Comment by stephanie-wang** [comment]
FYI, the load-balancing test that I added passes consistently on my laptop but not on Travis. I'm not sure exactly why, but it looks like it probably has something to do with the raylets ping-ponging waiting tasks, since they all think that another node has enough CPU resources to run the task. I'm reverting the test to the previous version that just checks whether it finishes instead of checking for even load. The other test that is added in this PR is good enough as a regression test.

**Comment by ericl** [comment]
LINT

**Comment by rkooo567** [comment]
`test_scheduling` seems to fail!

**Pull request merged**